perf(statusline): UI elements are always redrawn on K_EVENT

luukvbaal opened the pull request:
Problem:    'statusline'-format UI elements are redrawn on each K_EVENT.
Solution:   Only redraw UI elements when something relevant has changed.


Fix https://github.com/neovim/neovim/issues/4847 (untested)
Fix https://github.com/neovim/neovim/issues/14303
Fix https://github.com/neovim/neovim/issues/17937

luukvbaal commented:
Set a global variable to avoid redrawing the `'statusline'`-format UI elements. I won't pretend to understand all the implications of doing so (unfamiliar with the event loop). All tests still pass however after unsetting the boolean for cases where elements should explicitly be redrawn. Not sure if any untested cases are affected.

zeertzjq commented on src/nvim/ex_docmd.c:
```diff
@@ -6156,6 +6156,7 @@ static void ex_redrawstatus(exarg_T *eap)
     status_redraw_curbuf();
   }
 
+  skip_redr_custom_k_event = false;
```
This is also needed in `ex_redrawtabline()`

zeertzjq commented:
With this PR, sourcing the following Lua file causes the tabline to disappear:
```lua
vim.g.var = 'foo'
vim.o.showtabline = 2
vim.o.tabline = '%!g:var'
vim.schedule(function()
  vim.g.var = 'bar'
  vim.api.nvim_input('<C-L>')
end)
```

luukvbaal commented:
> With this PR, sourcing the following Lua file causes the tabline to disappear:

Fixed (by adding by unsetting the variable in the `<C-L>` code path). Does this seem like the correct approach if we can catch all the cases where we should ignore/unset `skip_redr_custom_k_event`?

luukvbaal commented:
> > With this PR, sourcing the following Lua file causes the tabline to disappear:

> Fixed (by unsetting the variable in the <C-L> code path).

The tabline is still cleared however (and not redrawn) for example when resizing the terminal. It seems to be cleared in `drawscreen.c:screenclear()` by `ui_call_grid_clear()`. 

I guess the tabline is part of the default grid and thus is cleared. Does it need to be cleared? Does it need to be part of the grid?

luukvbaal commented:
Previous comment might be irrelevant as it is fixed by the latest commit🤷🏻

justinmk commented:
Will be very nice to fix this, though adding another special-case global is unfortunate :)

justinmk commented on src/nvim/statusline.c:
```diff
@@ -279,7 +279,8 @@ static void win_redr_custom(win_T *wp, bool draw_winbar, bool draw_ruler)
   // Avoid trouble by not allowing recursion.
-  if (entered) {
+  if (entered || skip_redr_custom_k_event) {
+    skip_redr_custom_k_event = false;
```
Related globals `redraw_tabline` and `redraw_cmdline` look very similar in nature. Would it make sense to name this one `redraw_win_custom` ?

bfredl commented:
I don't think this is the right solution. There is nothing wrong with redrawing as a result of K_EVENT and this will_ break_ other plugins which use K_EVENT to respond directly to user intention (or respond external events in a proper fashion, not hot timer loop panic). 

Also we already have a set of variables to control _precisely_ what is going to be redrawn or not in the next redraw "tick". In particular, for a window we have `wp->w_redr_status` to control if the statusline is supposed to be redrawn. What we want is to avoid setting it to `true` in the first place. We need to find the precise code location that wants to set it to true in those linked issues and see if the condition can be made sharper so there is no redraw if no actual state has changed.

zeertzjq commented:
`show_cursor_info()` redraws custom statusline even if `w_redr_status` isn't set.

bfredl commented:
It is supposed only redraw the ruler or statusline when cursor was actually moved.

>  if false, only show ruler if position has changed.

But this only is actually checked for the builtin/default ruler or statusline. This check needs be moved and also apply when w_redr_custom redraws the ruler.

luukvbaal commented:
> This check needs be moved earlier in show_cursor_info()

Yeah that does seem like a better fix 👍🏻 Let me see if I can make that refactor without breaking any tests :)

luukvbaal committed:
test(statusline): UI elements are not redrawn on K_EVENT unnecessarily

luukvbaal commented:
> also, ideally, even this should be done as a check before update_screen() in normal_redraw() and ins_redraw() which sets the shared flag and then use the normal redraw entry point.

I briefly tried this but my attempt introduced too many test failures.

> This check needs be moved earlier in show_cursor_info()

Latest commit implements this. There are two remaining test failures that I can't really explain.

For some reason this change seems to break the `wordcount()` function🤡
And the "statusline is redrawn on entering cmdline with scrolled messages" on `test/functional/ui/cmdline_spec.lua:891` fails for some reason. The one that closes the scrolled message, the tests before it work fine. Don't really see how that matters for the code that handles this in `ex_getln.c:789`.

bfredl commented:
the failure in ` test/functional/ui/cmdline_spec.lua` is real. as we redraw the statusline when entering cmdline mode, thus redrawing it when exiting cmdline is reasonably to be expected. this was done implicitly before, but we need to do this explicitly now, maybe by setting `curwin->w_redr_status` in `ex_colon` but only if `getexline` was used.

luukvbaal commented:
> maybe by setting curwin->w_redr_status in ex_colon but only if getexline was used.

Adding a `status_redraw_all()` at the end of `command_line_enter()` fixed this (did that used to be called `ex_colon()`?). I don't think it needs to be done conditionally for `curwin` only? It is done unconditionally for all windows at the start of `command_line_enter()` as well.

Adding a `w_ru_state` that stores `get_real_state()` and checking for that in `show_cursor_info()` fixed the `wordcount()` test failure (those tests use `'statusline'` to store the wordcount in a variable so I guess the failing test was not that 🤡 after all).

I guess all these explicit statusline redraws that have to be added now reveals that `show_cursor_info()` redrawing unconditionally was responsible for updating the status line where it really shouldn't have been? I guess there might be other cases affected as well that have not been caught by the test-suite.

bfredl commented:
> I guess all these explicit statusline redraws that have to be added now reveals that show_cursor_info() unconditionally redrawing was responsible for updating the status line where it really shouldn't have been? I suppose there might be other cases affected as well that have not been caught by the test-suite.

Yes, this will likely affect some async plugins. But this will be more of an annoyance than a critical bugs. As issues get reported to us for these cases, we can then consider them individually which should be fixed in neovim, or in the plugin (by redrawing status explicit)

luukvbaal commented on src/nvim/ex_getln.c:
```diff
@@ -908,6 +908,7 @@ static uint8_t *command_line_enter(int firstc, long count, int indent, bool clea
     msg_ext_clear_later();
   }
 
+  status_redraw_all();  // redraw to show mode change
```
That was my initial instinct as well but it didn't resolve the test failure(`test/functional/ui/cmdline_spec.lua:891`).

luukvbaal replied:
Sorry, I lied. Seem to have remembered this wrong, it passes now.

luukvbaal committed:
perf(statusline): UI elements are always redrawn on K_EVENT

Problem:    'statusline'-format UI elements are redrawn on each K_EVENT.
Solution:   Only redraw UI elements when something relevant has changed.

luukvbaal commented:
> Yes, this will likely affect some async plugins. But this will be more of an annoyance than a critical bugs. As issues get reported to us for these cases, we can then consider them individually which should be fixed in neovim, or in the plugin (by redrawing status explicit)

Do you think we can still include this for 0.9 then? Perhaps accompanied with a `news.txt` item pointing out the change/workaround?

bfredl commented:
yes. The bug being fixed here is much more critical, so it will go in 0.9.

clason commented:
Also, any followup fixes can be backported; we usually put out a 0.x.1 bugfix release fairly quickly.

luukvbaal commented on src/nvim/drawscreen.c:
```diff
@@ -733,28 +734,53 @@ void show_cursor_info(bool always)
-      || (p_title && (stl_syntax & STL_IN_TITLE))) {
-    maketitle();
-  }
+    if (need_maketitle
```
Why are this and the `draw_tabline()` below conditional on `need_maketitle/redraw_tabline`? Either we want to support showing the cursor(now mode) information in the title/tabline or we don't. If we do they should be unconditionally called here I think, `set tal=%{mode()}` will not show the mode changes for example. Them being present in `show_cursor_info()` seems to indicate that we want to support this.

bfredl replied:
I agree it is weird, but let's consider that in a follow up PR (as I said, I want to refactor a lot of this code anyway)

bfredl replied:
although maybe the alternate `stl_syntax & STL_IN_TITLE)` condition handles this? too many magic variables..

luukvbaal replied:
Yeah for the title maybe. But like I said set `tal=%{mode()}` will not show mode changes due to the guard on `redraw_tabline` here.

Another weird thing: there is a `maketitle()` in `normal_redraw()` right before `show_cursor_info()` which will be redundant depending on what we do with this.
